ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Inbound.hs

```diff
@@ -0,0 +1,448 @@
+  , numToAckOnNextReq :: !NumObjectIdsAck
+  -- ^ The number of objects we can acknowledge on our next request
+  -- for more object IDs. Their corresponding IDs have already been removed
+  -- from 'outstandingFifo'.
+  }
+  deriving stock (Show, Generic)
+  deriving anyclass NoThunks
+
+initialInboundSt :: InboundSt objectId object
+initialInboundSt = InboundSt 0 Seq.empty Set.empty Map.empty 0
+
+objectDiffusionInbound ::
+  forall objectId object m.
+  ( Ord objectId
+  , NoThunks objectId
+  , NoThunks object
+  , MonadSTM m
+  , MonadThrow m
+  ) =>
+  Tracer m (TraceObjectDiffusionInbound objectId object) ->
+  -- | Maximum values for outstanding FIFO length, number of IDs to request,
+  -- and number of objects to request
+  (NumObjectsOutstanding, NumObjectIdsReq, NumObjectsReq) ->
+  ObjectPoolWriter objectId object m ->
+  NodeToNodeVersion ->
+  ControlMessageSTM m ->
+  ObjectDiffusionInboundPipelined objectId object m ()
+objectDiffusionInbound
+  tracer
+  (maxFifoLength, maxNumIdsToReq, maxNumObjectsToReq)
+  ObjectPoolWriter{..}
+  _version
+  controlMessageSTM =
+    ObjectDiffusionInboundPipelined $!
+      checkState initialInboundSt & go Zero
+   where
+    canRequestMoreObjects :: InboundSt k object -> Bool
+    canRequestMoreObjects st =
+      not (Set.null (canRequestNext st))
+
+    -- Computes how many new IDs we can request so that receiving all of them
+    -- won't make 'outstandingFifo' exceed 'maxFifoLength'.
+    numIdsToReq :: InboundSt objectId object -> NumObjectIdsReq
+    numIdsToReq st =
+      maxNumIdsToReq
+        `min` ( fromIntegral maxFifoLength
+                  - (fromIntegral $ Seq.length $ outstandingFifo st)
+                  - numIdsInFlight st
+              )
+
+    -- Updates 'InboundSt' with new object IDs and return the updated 'InboundSt'.
+    --
+    -- Collected object IDs that are already in the objectPool are pre-emptively
+    -- acknowledged so that we don't need to bother requesting them from the
+    -- outbound peer.
+    preAcknowledge ::
+      InboundSt objectId object ->
+      (objectId -> Bool) ->
+      [objectId] ->
+      InboundSt objectId object
+    preAcknowledge st _ collectedIds | null collectedIds = st
+    preAcknowledge st poolHasObject collectedIds =
+      let
+        -- Divide the collected IDs in two parts: those that are already in the
+        -- objectPool and those that are not.
+        (alreadyObtained, notYetObtained) =
+          List.partition
+            (\objectId -> poolHasObject objectId)
+            collectedIds
+
+        -- The objects that we intentionally don't request, because they are
+        -- already in the objectPool, will need to be acknowledged.
+        -- So we extend 'pendingObjects' with those objects (so of course they
+        -- have no corresponding reply).
+        pendingObjects' =
+          pendingObjects st
+            <> Map.fromList [(objectId, Nothing) | objectId <- alreadyObtained]
```

Comment:
NIT: depending on the size of both `pendingObjects st` and `alreadyObtained`, it might make sense to turn this into a fold, e.g.:

```haskell
pendingObjects' =
  Map.foldl' 
    (flip Map.insert Nothing) 
    (pendingObjects st) 
    alreadyObtained
```

This would also match how we delete multiple objects later on L200.